### ISSUE TITLE
NULL pointer dereference in DohSeek

### ISSUE THREAD

**Issue by gy741** [swig/swig#1172]
Hello.

I found a NULL pointer dereference bug in swig.

Please confirm.

PoC Download: [Null_DohSeek.zip](https://github.com/swig/swig/files/1569503/Null_DohSeek.zip)

Thanks.

OS: CentOS 7 64bit
Version: swig-3.0.12
Steps to reproduce:
1.Download the .POC files.
2.Compile the source code with ASan.
3.Execute the following command
: ./swig $POC

```
=================================================================
==11184==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x7ff7ec07060c bp 0x7ffcdc9c90b0 sp 0x7ffcdc9c8ff0 T0)
==11184==The signal is caused by a READ memory access.
==11184==Hint: address points to the zero page.
    #0 0x7ff7ec07060b in __GI_fseek (/lib64/libc.so.6+0x7360b)
    #1 0x5f573b in DohSeek /home/karas/swig-3.0.12/Source/DOH/base.c:674:10
    #2 0xb55135 in retract /home/karas/swig-3.0.12/Source/Swig/scanner.c:382:11
    #3 0xb4b05f in look /home/karas/swig-3.0.12/Source/Swig/scanner.c:757:6
    #4 0xb46be1 in Scanner_token /home/karas/swig-3.0.12/Source/Swig/scanner.c:1407:7
    #5 0x51e75e in yylook /home/karas/swig-3.0.12/Source/CParse/cscanner.c:195:16
    #6 0x5199f0 in yylex /home/karas/swig-3.0.12/Source/CParse/cscanner.c:445:7
    #7 0x51ff4b in yyparse /home/karas/swig-3.0.12/Source/y.tab.c:4400:16
    #8 0x51f445 in Swig_cparse /home/karas/swig-3.0.12/Source/parser.y:1099:3
    #9 0x9094ec in SWIG_main(int, char**, Language*) /home/karas/swig-3.0.12/Source/Modules/main.cxx:1178:17
    #10 0xae5f49 in main /home/karas/swig-3.0.12/Source/Modules/swigmain.cxx:207:13
    #11 0x7ff7ec01ec04 in __libc_start_main (/lib64/libc.so.6+0x21c04)
    #12 0x41c157 in _start (/home/karas/swig-3.0.12/swig+0x41c157)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (/lib64/libc.so.6+0x7360b) in __GI_fseek
==11184==ABORTING
```

==========================================
[Acknowledgement]
This work was supported by ICT R&D program of MSIP/IITP. [R7518-16-1001, Innovation hub for high Performance Computing]

**Comment by wsfulton** [comment]
I can't reproduce. I don't really follow your instructions though. I unzipped Doh_NullSeek.zip and ran:
```
swig -c++  /tmp/issue1172/Null_DohSeek
/tmp/issue1172/Null_DohSeek:5: Error: Unterminated raw string, started with R"

  Some docstrings.



Different indent.

";



%inline %{

  xxx( is not terminated by )

  Some docstrings.



Different indent.

";



%inline %{

  xxx"
/tmp/issue1172/Null_DohSeek:5: Error: Syntax error in input(1).
```
which gives a technically correct error. How did you build SWIG with Asan? I tried using gcc and valgrind but it didn't show up any problems.

**Comment by wsfulton** [comment]
I used latest on master which gives the error shown above. I couldn't replicate it as this bug was fixed in dee6b075a83468bacb39a1d9970d2c65f3e68d9b after swig-3.0.12 was released. Nevertheless, thanks for the bug report.

I also managed to replicate the AddressSanitizer error in swig-3.0.12 configuring using:
```
./configure CXXFLAGS="-O1 -g -fsanitize=address -fno-omit-frame-pointer" CFLAGS="-O1 -g -fsanitize=address -fno-omit-frame-pointer" CC=clang CXX=clang++
```
I ran `make partialcheck-test-suite` with this setup and it found some other errors which I'll fix, so it was a useful exercise.

**Comment by gy741** [comment]
@wsfulton 

Hello,

I checked the patch.

But I found a new bug.

Please confirm.

Thanks.

OS: CentOS 7 64bit
Version: commit 368cd3b52c8a17dd829ac42481a1901b449f9a60
PoC Download:  [new_null_error.zip](https://github.com/swig/swig/files/1628434/new_null_error.zip)

Steps to reproduce:
1.Download the .POC files.
2.Compile the source code with ASan.
3.Execute the following command
: ./swig $POC

```
ASAN:DEADLYSIGNAL
=================================================================
==2835==ERROR: AddressSanitizer: SEGV on unknown address 0x00000000 (pc 0xb7af2947 bp 0xbf8454a8 sp 0xbf845490 T0)
    #0 0xb7af2946 in fseek /build/glibc-KM3i_a/glibc-2.23/libio/fseek.c:32
    #1 0x819f7ea in DohSeek /root/karas/swig/Source/DOH/base.c:674:10
    #2 0x84161b4 in retract /root/karas/swig/Source/Swig/scanner.c:382:11
    #3 0x84118e5 in look /root/karas/swig/Source/Swig/scanner.c:1121:2
    #4 0x84118e5 in Scanner_token /root/karas/swig/Source/Swig/scanner.c:1411
    #5 0x83e9389 in expr_token /root/karas/swig/Source/Preprocessor/expr.c:242:9
    #6 0x83e9389 in Preprocessor_expr /root/karas/swig/Source/Preprocessor/expr.c:278
    #7 0x83e2b6c in Preprocessor_parse /root/karas/swig/Source/Preprocessor/cpp.c:1657:12
    #8 0x83e2318 in Preprocessor_parse /root/karas/swig/Source/Preprocessor/cpp.c:1908:13
    #9 0x82e001c in SWIG_main(int, char**, Language*) /root/karas/swig/Source/Modules/main.cxx:1084:9
    #10 0x83c6abd in main /root/karas/swig/Source/Modules/swigmain.cxx:207:13
    #11 0xb7aa5636 in __libc_start_main /build/glibc-KM3i_a/glibc-2.23/csu/../csu/libc-start.c:291
    #12 0x807862c in _start (/root/karas/swig/run/bin/swig+0x807862c)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /build/glibc-KM3i_a/glibc-2.23/libio/fseek.c:32 in fseek
==2835==ABORTING
```

==========================================
[Acknowledgement]
This work was supported by ICT R&D program of MSIP/IITP. [R7518-16-1001, Innovation hub for high Performance Computing]

**Comment by wsfulton** [comment]
I fixed this second one in 07a30249f416ced2ae627e1a32c15d38cb403087.
